proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,222 @@
+
+### Pitfalls of the current solution
+
+* We don't have an officially documented migration strategy for WAL and TSDB in general (only tribal knowledge).
+* Users don't know if version X of Prometheus works with version X-2 (except on some minor release that makes the change).
+* No e2e testing for migration strategies, no consistent documentation when this happens on releases.
+
+As a result:
+
+* Reduced contribution and development velocity, prolonging bugs, important features and efficiency efforts.
+* Increased operational cost and reduced trust for users who run highly durable Prometheus setups.
+
+### Context: TSDB format changes
+
+TSDB format, including chunk format (also used in the head with mmapped chunks), is also a handwritten variadic-encoding binary format. Compared to WAL it's generally versioned e.g.
+
+* [Chunk disk file format](https://github.com/prometheus/prometheus/blob/12c39d5421cc29a4bfc13fc57fd9ccd3dbc310f0/tsdb/docs/format/chunks.md#L14).
+* [Head chunk disk format](https://github.com/prometheus/prometheus/blob/fd62dbc2918deea0ceae94758baf7a095b52dd5b/tsdb/docs/format/head_chunks.md#L13).
+* [Index file](https://github.com/prometheus/prometheus/blob/d699dc3c7706944aafa56682ede765398f925ef0/tsdb/docs/format/index.md#L8).
+* [Block meta file](https://github.com/prometheus/prometheus/blob/5e124cf4f2b9467e4ae1c679840005e727efd599/tsdb/block.go#L171).
+* Chunks have its own ["encoding type"](https://github.com/prometheus/prometheus/blob/a693dd19f244f000da40bcbac85041846b78cfc1/tsdb/chunkenc/chunk.go#L29) that could be used for versioning.
+
+No docs or formal strategy was developed, but Prometheus generally follow the ["2-fold" migration strategy explained below](#two-fold-migration-strategy).
+
+## Goals
+
+* Agree on the official recommendations for **lossless WAL migrations** strategies for devs and users.
+* Balancing development velocity with user data stability risks.
+
+## Non-Goals
+
+* TSDB format can but does not need to follow the same recommendations. We want to change the WAL format now, so prioritizing WAL.
+* Mentioning Write-Before-Log, it uses WAL format so all apply to WBL too.
+* To reduce scope we don't mention [memory snapshot format](https://github.com/prometheus/prometheus/blob/fd5ea4e0b590d57df1e2ef41d027f2a4f640a46c/tsdb/docs/format/memory_snapshot.md#L1) for now.
+
+## How
+
+We recommend the [Two-Fold Migration Strategy](#two-fold-migration-strategy) with one variation: a configuration flag that tells Prometheus to write a Y+1 version if the user desires.
+
+
+In other words, we propose to add a TSDB **integer** flag `--storage.tsdb.write-wal-format` that tells Prometheus to use a particular WAL format for both WAL and WBL. This kind of flag will change its default to a new version ONLY when the previous Prometheus version can read that version.
```

Comment:
I think we should be more specific about the version here. Prometheus defines LTS versions, so this should read 
```suggestion
In other words, we propose to add a TSDB **integer** flag `--storage.tsdb.write-wal-format` that tells Prometheus to use a particular WAL format for both WAL and WBL. This kind of flag will change its default to a new version ONLY when the previous LTS Prometheus version can read that version.
```

This would mean 2 years life time of a WAL version (but of course it may mean a jump of more than one version). So let's suppose we implement this flag and the default is  1 in version 3.1 LTS (or something) . To set the default to 2 (or more) we need an LTS release a year later that can read version 2, but defaults to 1. Another year later we could set default 2.

WDYT?

Reply:
oh, I later read the alternatives - I think it makes sense to instill a sense of stability

Reply:
Ah, sorry. Copy-pasta. I really wanted to NOT do LTS for now, but initially I thought that's a good idea. This should be in alternatives. WDYT?

Reply:
Yes, one year https://prometheus.io/docs/introduction/release-cycle/

Reply:
Yes, this would extend what we said we would do in LTS releases. 
The risk is something to weigh up.